http/src/v0/refs/options.rs

```diff
@@ -0,0 +1,143 @@
+    pub unique: bool,
+    pub recursive: bool,
+    // `int` in the docs apparently is platform specific
+    // go-ipfs only honors this when `recursive` is true.
+    // go-ipfs treats -2 as -1 when `recursive` is true.
+    // go-ipfs doesn't use the json return value if this value is too large or non-int
+    pub max_depth: Option<i64>,
+}
+
+impl RefsOptions {
+    pub fn max_depth(&self) -> Option<u64> {
+        if self.recursive {
+            match self.max_depth {
+                // zero means do nothing
+                Some(x) if x >= 0 => Some(x as u64),
+                _ => None,
+            }
+        } else {
+            // only immediate links after the path
+            Some(1)
+        }
+    }
+}
+
+#[derive(Debug)]
+pub enum ParseError<'a> {
+    DuplicateField(Cow<'a, str>),
+    MissingArg,
+    InvalidNumber(Cow<'a, str>, Cow<'a, str>),
+    InvalidBoolean(Cow<'a, str>, Cow<'a, str>),
+}
+
+impl<'a> fmt::Display for ParseError<'a> {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        use ParseError::*;
+        match *self {
+            DuplicateField(ref s) => write!(fmt, "field {:?} was duplicated", *s),
+            MissingArg => write!(fmt, "required field \"arg\" missing"),
+            InvalidNumber(ref k, ref v) => write!(fmt, "field {:?} invalid number: {:?}", *k, *v),
+            InvalidBoolean(ref k, ref v) => write!(fmt, "field {:?} invalid boolean: {:?}", *k, *v),
+        }
+    }
+}
+
+impl<'a> std::error::Error for ParseError<'a> {}
+
+impl<'a> TryFrom<&'a str> for RefsOptions {
+    type Error = ParseError<'a>;
+
+    fn try_from(q: &'a str) -> Result<Self, Self::Error> {
+        use ParseError::*;
+
+        // TODO: check how go-ipfs handles duplicate parameters for non-Vec fields
+        //
+        // this manual deserialization is required because `serde_urlencoded` (used by
+        // warp::query) does not support multiple instances of the same field, nor does
+        // `serde_qs` (it would support arg[]=...). supporting this in `serde_urlencoded` is
+        // out of scope; not sure of `serde_qs`.
+        let parse = url::form_urlencoded::parse(q.as_bytes());
+
+        let mut args = Vec::new();
+        let mut format = None;
+        let mut edges = None;
+        let mut unique = None;
+        let mut recursive = None;
+        let mut max_depth = None;
+
+        for (key, value) in parse {
+            let target = match &*key {
+                "arg" => {
+                    args.push(value.into_owned());
+                    continue;
+                }
+                "format" => {
+                    if format.is_none() {
+                        // not parsing this the whole way as there might be hope to have this
+                        // function removed in the future.
+                        format = Some(value.into_owned());
+                        continue;
+                    } else {
+                        return Err(DuplicateField(key));
```

Comment:
will this error if a format arg is passed at all?

Reply:
Nope, only on the second `format=foo&format=bar`